23-rseq/rseq.2
.\" Copyright 2015-2020 Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
.\"
.\" %%%LICENSE_START(VERBATIM)
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\"
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\"
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\" %%%LICENSE_END
.\"
.TH RSEQ 2 2020-06-05 "Linux" "Linux Programmer's Manual"
.SH NAME
rseq \- Restartable sequences and cpu number cache
.SH SYNOPSIS
.nf
.B #include <linux/rseq.h>
.sp
.BI "int rseq(struct rseq * " rseq ", uint32_t " rseq_len ", int " flags ", uint32_t " sig ");
.sp
.SH DESCRIPTION

A restartable sequence is a sequence of instructions guaranteed to be executed
atomically with respect to other threads and signal handlers on the current
CPU. If its execution does not complete atomically, the kernel changes the
execution flow by jumping to an abort handler defined by user-space for that
restartable sequence.

Using restartable sequences requires to register a
.BR __rseq_abi
thread-local storage data structure (struct rseq) through the
.BR rseq ()
system call. Only one
.BR __rseq_abi
can be registered per thread, so user-space libraries and applications must
follow a user-space ABI defining how to share this resource.  The ABI defining
how to share this resource between applications and libraries is defined by the
C library.

The
.BR __rseq_abi
contains a
.I rseq_cs
field which points to the currently executing critical section. For each
thread, a single rseq critical section can run at any given point. Each
critical section need to be implemented in assembly.

The
.BR rseq ()
ABI accelerates user-space operations on per-cpu data by defining a
shared data structure ABI between each user-space thread and the kernel.

It allows user-space to perform update operations on per-cpu data
without requiring heavy-weight atomic operations.

The term CPU used in this documentation refers to a hardware execution
context. For instance, each CPU number returned by
.BR sched_getcpu ()
is a CPU. The current CPU means to the CPU on which the registered thread is
running.

Restartable sequences are atomic with respect to preemption (making it
atomic with respect to other threads running on the same CPU), as well
as signal delivery (user-space execution contexts nested over the same
thread). They either complete atomically with respect to preemption on
the current CPU and signal delivery, or they are aborted.

Restartable sequences are suited for update operations on per-cpu data.

Restartable sequences can be used on data structures shared between threads
within a process, and on data structures shared between threads across
different processes.

.PP
Some examples of operations that can be accelerated or improved
by this ABI:
.IP \[bu] 2
Memory allocator per-cpu free-lists,
.IP \[bu] 2
Querying the current CPU number,
.IP \[bu] 2
Incrementing per-CPU counters,
.IP \[bu] 2
Modifying data protected by per-CPU spinlocks,
.IP \[bu] 2
Inserting/removing elements in per-CPU linked-lists,
.IP \[bu] 2
Writing/reading per-CPU ring buffers content.
.IP \[bu] 2
Accurately reading performance monitoring unit counters
with respect to thread migration.

.PP
Restartable sequences must not perform system calls. Doing so may result
in termination of the process by a segmentation fault.

.PP
The
.I rseq
argument is a pointer to the thread-local rseq structure to be shared
between kernel and user-space.

.PP
The layout of
.B struct rseq
is as follows:
.TP
.B Structure alignment
This structure is aligned on 32-byte boundary.
.TP
.B Structure size
This structure is fixed-size (32 bytes). Its size is passed as parameter to the
rseq system call.
.PP
.in +8n
.EX
struct rseq {
    __u32 cpu_id_start;
    __u32 cpu_id;
    union {
        /* Edited out for conciseness. [...] */
    } rseq_cs;
    __u32 flags;
} __attribute__((aligned(32)));
.EE
.TP
.B Fields

.TP
.in +4n
.I cpu_id_start
Optimistic cache of the CPU number on which the registered thread is
running. Its value is guaranteed to always be a possible CPU number,
even when rseq is not registered. Its value should always be confirmed by
reading the cpu_id field before user-space performs any side-effect (e.g.
storing to memory).

This field is an optimistic cache in the sense that it is always
guaranteed to hold a valid CPU number in the range [ 0 ..
nr_possible_cpus - 1 ]. It can therefore be loaded by user-space and
used as an offset in per-cpu data structures without having to
check whether its value is within the valid bounds compared to the
number of possible CPUs in the system.

Initialized by user-space to a possible CPU number (e.g., 0), updated
by the kernel for threads registered with rseq.

For user-space applications executed on a kernel without rseq support,
the cpu_id_start field stays initialized at 0, which is indeed a valid
CPU number. It is therefore valid to use it as an offset in per-cpu data
structures, and only validate whether it's actually the current CPU
number by comparing it with the cpu_id field within the rseq critical
section. If the kernel does not provide rseq support, that cpu_id field
stays initialized at -1, so the comparison always fails, as intended.

It is up to user-space to implement a fall-back mechanism for scenarios where
rseq is not available.
.in
.TP
.in +4n
.I cpu_id
Cache of the CPU number on which the registered thread is running. Initialized
by user-space to -1, updated by the kernel for threads registered with rseq.
.in
.TP
.in +4n
.I rseq_cs
The rseq_cs field is a pointer to a struct rseq_cs. Is is NULL when no
rseq assembly block critical section is active for the registered thread.
Setting it to point to a critical section descriptor (struct rseq_cs)
marks the beginning of the critical section.

Initialized by user-space to NULL.

Updated by user-space, which sets the address of the currently
active rseq_cs at the beginning of assembly instruction sequence
block, and set to NULL by the kernel when it restarts an assembly
instruction sequence block, as well as when the kernel detects that
it is preempting or delivering a signal outside of the range
targeted by the rseq_cs. Also needs to be set to NULL by user-space
before reclaiming memory that contains the targeted struct rseq_cs.

Read and set by the kernel.
.in
.TP
.in +4n
.I flags
Flags indicating the restart behavior for the registered thread. This is
mainly used for debugging purposes. Can be a combination of:
.IP \[bu]
RSEQ_CS_FLAG_NO_RESTART_ON_PREEMPT: Inhibit instruction sequence block restart
on preemption for this thread.
.IP \[bu]
RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL: Inhibit instruction sequence block restart
on signal delivery for this thread.
.IP \[bu]
RSEQ_CS_FLAG_NO_RESTART_ON_MIGRATE: Inhibit instruction sequence block restart
on migration for this thread.
.in

Initialized by user-space, used by the kernel.

.PP
The layout of
.B struct rseq_cs
version 0 is as follows:
.TP
.B Structure alignment
This structure is aligned on 32-byte boundary.
.TP
.B Structure size
This structure has a fixed size of 32 bytes.
.PP
.in +8n
.EX
struct rseq_cs {
    __u32   version;
    __u32   flags;
    __u64   start_ip;
    __u64   post_commit_offset;
    __u64   abort_ip;
} __attribute__((aligned(32)));
.EE
.TP
.B Fields

.TP
.in +4n
.I version
Version of this structure. Should be initialized to 0.
.in
.TP
.in +4n
.I flags
Flags indicating the restart behavior of this structure. Can be a combination
of:
.IP \[bu]
RSEQ_CS_FLAG_NO_RESTART_ON_PREEMPT: Inhibit instruction sequence block restart
on preemption for this critical section.
.IP \[bu]
RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL: Inhibit instruction sequence block restart
on signal delivery for this critical section.
.IP \[bu]
RSEQ_CS_FLAG_NO_RESTART_ON_MIGRATE: Inhibit instruction sequence block restart
on migration for this critical section.
.TP
.in +4n
.I start_ip
Instruction pointer address of the first instruction of the sequence of
consecutive assembly instructions.
.in
.TP
.in +4n
.I post_commit_offset
Offset (from start_ip address) of the address after the last instruction
of the sequence of consecutive assembly instructions.
.in
.TP
.in +4n
.I abort_ip
Instruction pointer address where to move the execution flow in case of
abort of the sequence of consecutive assembly instructions.
.in

.PP
The
.I rseq_len
argument is the size of the
.I struct rseq
to register.

.PP
The
.I flags
argument is 0 for registration, and
.IR RSEQ_FLAG_UNREGISTER
for unregistration.

.PP
The
.I sig
argument is the 32-bit signature to be expected before the abort
handler code.

.PP
A single library per process should keep the rseq structure in a
thread-local storage variable.
The
.I cpu_id
field should be initialized to -1, and the
.I cpu_id_start
field should be initialized to a possible CPU value (typically 0).

.PP
Each thread is responsible for registering and unregistering its rseq
structure. No more than one rseq structure address can be registered
per thread at a given time.

.PP
Reclaim of rseq object's memory must only be done after either an
explicit rseq unregistration is performed or after the thread exits.

.PP
In a typical usage scenario, the thread registering the rseq
structure will be performing loads and stores from/to that structure. It
is however also allowed to read that structure from other threads.
The rseq field updates performed by the kernel provide relaxed atomicity
semantics (atomic store, without memory ordering), which guarantee that other
threads performing relaxed atomic reads (atomic load, without memory ordering)
of the cpu number cache will always observe a consistent value.

.SH RETURN VALUE
A return value of 0 indicates success. On error, \-1 is returned, and
.I errno
is set appropriately.

.SH ERRORS
.TP
.B EINVAL
Either
.I flags
contains an invalid value, or
.I rseq
contains an address which is not appropriately aligned, or
.I rseq_len
contains an incorrect size.
.TP
.B ENOSYS
The
.BR rseq ()
system call is not implemented by this kernel.
.TP
.B EFAULT
.I rseq
is an invalid address.
.TP
.B EBUSY
Restartable sequence is already registered for this thread.
.TP
.B EPERM
The
.I sig
argument on unregistration does not match the signature received
on registration.

.SH VERSIONS
The
.BR rseq ()
system call was added in Linux 4.18.

.SH CONFORMING TO
.BR rseq ()
is Linux-specific.

.in
.SH SEE ALSO
.BR sched_getcpu (3) ,
.BR membarrier (2)

23-rseq/rseq.c
#define _GNU_SOURCE
#include <pthread.h>
#include <unistd.h>
#include <inttypes.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <sys/syscall.h>
#include <stdatomic.h>
#include <string.h>
#include <sys/sysinfo.h>
#include <stdbool.h>
#include <malloc.h>

#define die(msg) do { perror(msg); exit(EXIT_FAILURE); } while(0)

// With modern glibc versions (>2.35), the glibc already registers
// a rseq area for each thread that is started with
// pthread_create. In that case, we derive this pointer from the
// thread_pointer(). For details, see
// https://www.gnu.org/software/libc/manual/html_node/Restartable-Sequences.html
#if __has_include ("sys/rseq.h")
#include <sys/rseq.h>

#if RSEQ_SIG != 0x53053053
#error "glibc defined RSEQ_SIG differently"
#endif

static struct rseq * rseq_register() {
    return __builtin_thread_pointer() + __rseq_offset;
}

#else
#include <linux/rseq.h>
#define RSEQ_SIG   0x53053053

// The rseq(2) syscall has no glibc wrapper. Therefore, we define our
// own. Please run `make man` to see the manpage rseq(2).
int sys_rseq(struct rseq * rseq, uint32_t rseq_len, int flags, uint32_t sig) {
    return syscall(SYS_rseq, rseq, rseq_len, flags, RSEQ_SIG);
}

struct rseq *rseq_register() {
    struct rseq *ret = memalign(sizeof(struct rseq), sizeof(struct rseq));
    memset(ret, 0, sizeof(struct rseq));
    ret->cpu_id_start = -1;
    ret->cpu_id       = -1;
    if (sys_rseq(ret, sizeof(struct rseq), 0, 0) < 0)
        die("rseq");
    return ret;
}

#endif



// This data structure is exactly one cache-line wide (assuming that a
// cache line is 64 bytes). Thereby, we can allocate an array of
// cpu-local counters, where each CPU only operates on a single cache
// line. Thereby, we can avoid most side effects of cache-line
// transfers.
struct cacheline {
    union {
        char data[64];
        struct {
            uint64_t        counter;
            pthread_mutex_t mutex;   // Used in the lock variant
        };
    };
};

// We will define multiple operation_t functions that all implement
// the same behavior: They increment a cpu-local counter by 1.
typedef int (*operation_t)(struct rseq *_, struct cacheline *);

// The simplest variant of a cpu-local counter is to get the cpuid
// with getcpu() and increment the counter. However, due to the
// read-update-write cycle, this variant is racy and will produce
// incorrect results.
int operation_regular(struct rseq*_, struct cacheline *counters) {
    unsigned int cpu_id;

    getcpu(&cpu_id, NULL);
    counters[cpu_id].counter += 1;

    return 0;
}

// A correct, but slow variant uses the cache-line--local pthread
// mutex to lock the counter for the time of the operation.
int operation_lock(struct rseq*_, struct cacheline *counters) {
    unsigned int cpu_id;

    getcpu(&cpu_id, NULL);

    pthread_mutex_lock(&counters[cpu_id].mutex);
    counters[cpu_id].counter += 1;
    pthread_mutex_unlock(&counters[cpu_id].mutex);

    return 0;
}

// Variant that uses getcpu() + atomic_fetch_add
int operation_atomic(struct rseq* _, struct cacheline *counters) {
    // FIXME: Implement variant

    return 0;
}

// Variant without getcpu: Like operation_atomic, but uses the
// restartable sequence to retrieve the cpu id.
// Please look at /usr/include/linux/rseq.h for the documentation of struct rseq
int operation_rseq_atomic(struct rseq* rs, struct cacheline *counters) {
    // FIXME: Implement variant

    return 0;
}


// Variant that uses no atomic operations and fully relies on rseq
// This variant is implemented in assembler (see rseq.S)
extern int operation_rseq(struct rseq *, struct cacheline*);
// FIXME: Implement counter_rseq in rseq.S


////////////////////////////////////////////////////////////////
// The Benchmarking code
//
// We start NTHREADS threads and each thread executes
// ROUNDS_PER_THREAD cpu-local increments

int  ROUNDS_PER_THREAD = 50000000;

struct thread_args {
    operation_t         operation;
    struct cacheline    *counters;
};


void* thread_handler(void* data) {
    struct thread_args *args = data;

    // Register rseq area or use glibc's rseq
    struct rseq *rseq = rseq_register();
    printf("rseq: %p\n", rseq);
    
    // Execute the given operation ROUNDS_PER_THREAD times and count
    // the number of aborts (only != 0 for rseq)
    uint64_t aborts = 0;
    for (uint64_t i = 0; i < ROUNDS_PER_THREAD; i++) {
        aborts += args->operation(rseq, args->counters);
    }

    // Return the number of rseq aborts
    return (void*) aborts;
}

// Print usage and exit.
static void usage(char *argv0) {
    fprintf(stderr, "usage: %s <threads> <regular|lock|getcpu-atomic|rseq-atomic|rseq> [rounds]\n", argv0);
    exit(EXIT_FAILURE);
}

int main(int argc, char *argv[]) {
    // Parameter Parsing. This is boring
    if (argc < 3) usage(argv[0]);
    if (argc == 4)
        ROUNDS_PER_THREAD *= atoi(argv[3]);

    int CPUS     = get_nprocs();
    int NTHREADS = atoi(argv[1]);
    char *MODE   = argv[2];

    struct thread_args args;
    if      (!strcmp(MODE, "rseq"))          args.operation = operation_rseq;
    else if (!strcmp(MODE, "getcpu-atomic")) args.operation = operation_atomic;
    else if (!strcmp(MODE, "rseq-atomic"))   args.operation = operation_rseq_atomic;
    else if (!strcmp(MODE, "regular"))       args.operation = operation_regular;
    else if (!strcmp(MODE, "lock"))          args.operation = operation_lock;
    else      usage(argv[0]);

    // Initialize the cpu-local counters. Each CPU gets an struct
    // cacheline on its own. We use aligned_alloc(3) to get
    // cacheline-aligned memory from the allocator

    args.counters = aligned_alloc(sizeof(struct cacheline), CPUS * sizeof(struct cacheline));
    if (!args.counters) die("calloc");


    // Initialize locks for the lock variant
    for (uint32_t i = 0; i < CPUS; i++) {
        pthread_mutex_init(&args.counters[i].mutex, NULL);
    }


    // The actual benchmarking code
    ////////////////////////////////////////////////////////////////
    struct timespec start, end;
    // Start Time. We use the CLOCK_PROCESS_CPUTIME_ID to get the
    // number of cpu seconds spent. 
    if (clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &start) < 0)
        die("clock_gettime");

    // Create NTHREADS threads
    pthread_t threads[NTHREADS];
    for (uint32_t i = 0; i < NTHREADS; i++) {
        pthread_create(&threads[i], NULL, thread_handler, (void*)&args);
    }

    // Wait for all threads to complete and accumulate the number of aborts
    uint64_t aborts = 0;
    for (uint32_t i = 0; i < NTHREADS; i++) {
        uint64_t thread_aborts;
        pthread_join(threads[i], (void**)&thread_aborts);
        aborts += thread_aborts;
    }

    // End Time
    if (clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &end) < 0)
        die("clock_gettime");

    // Calculate the time delta between both points in time.
    double delta = end.tv_sec - start.tv_sec;
    delta += (end.tv_nsec - start.tv_nsec) / 1e9;

    // Print out the cpu-local counters. With this output and a low
    // number of threads you can see the thread migration.
    uint64_t sum = 0;
    for (uint32_t i = 0; i < CPUS; i++) {
        fprintf(stderr, "counter[cpu=%d] = %ld\n", i, args.counters[i].counter);
        sum += args.counters[i].counter;
    }

    // Print out the result. We also check that the threads actually
    // counted correctly (state)
    printf("mode=%s threads=%d sum=%ld state=%s aborts=%ld cputime=%fs per_increment=%fns\n",
           MODE, NTHREADS,
           sum, (sum % ROUNDS_PER_THREAD) == 0 ? "ok" : "fail",
           aborts,
           delta,            // total cpu time that was spent
           delta * 1e9 / sum // nanoseconds per increment
        );
}
